# Ukázka práce s otevřenými daty o úkonech z Registru práv a povinností

Otevřená data z registru práv a povinností (RPP) sestávají z referenčních údajů o působnostech jednotlivých orgánů veřejné moci (OVM). Mezi orgány veřejné moci se řadí instituce veřejné správy jako např. centrální úřady (např. Česká správa sociálního zabezpečení), ministerstva, obce (resp. jejich obecní úřady), ale např. i vysoké školy.

Součástí údajů o působnostech OVM vedených v RPP jsou údaje o úkonech, které OVM vykonávají. Úkony odpovídají ve většině případů službám veřejné správy, kdy konkrétní úředník umožňuje nebo pomáhá konkrétnímu občanovi vypořádat jeho práva nebo povinnosti. Příkladem úkonu je např. "Ohlašení stavebního záměru", "Vydání stavebního povolení", "Vydání osvědčení o rovnocenosti vzdělání do zahraničí", "Žádost o registraci klenotnické slitiny nebo její změnu" nebo "Vydání cestovního pasu".

Otevřená data z RPP umožňují vytvořit aplikaci, která by podávala přehled úkonů vykonávaných veřejnou správou, ukazovala, na jakých úřadech může občan, podnikatel nebo právnická osoba vybraný úkon vyřídit, doporučovala související úkony nebo informovala, zda lze daný úkon vyřešit i elektronicky. Dokonce RPP ve spojení s RÚIAN (Registr územních identifikátorů, adres a nemovitostí) umožňuje zobrazovat na mapě ČR, na jakých pracovištích lze dané úkony řešit a existují i data, která obsahují otevírací doby těchto pracovišť.

Takto načrtnutá aplikace ještě neexistuje, už jenom z toho důvodu, že data o úkonech v RPP jsou poměrně čerstvá. Vzhledem k jejich čerstvosti jsou také ale v některých oblastech neúplná. Např. Úřad práce eviduje v RPP pouze jedno pracoviště a tak z RPP vyplývá, že úkony týkající se sociálních dávek musí občané řešit pouze na jednom pracovišti v Praze, což je evidentně nesmysl. Další přidanou hodnotou aplikace by tak mohlo být poukazování na tyto neúplnosti, což může být zajímavou zpětnou vazbou pro úřady, které data do RPP dodávají, nebo pro NKÚ, který veřejnou správu kontroluje.

Zdrojový kód tohoto Jupyter Notebooku lze [stáhnout](https://github.com/opendata-mvcr/opendata-mvcr.github.io/blob/master/datové-články/registr-práv-a-povinností/rpp-úkony.ipynb) z GitHubu. Nezapomeňte na [obrázky s konceptuálními schématy](https://github.com/opendata-mvcr/opendata-mvcr.github.io/tree/master/datové-články/registr-práv-a-povinností/img).

## Konceptuální schéma úkonů v RPP

Na úrovni konceptuálního modelu RPP jsou úkony instancemi třídy [Úkon](https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/úkon). Kompletní konceptuální schéma úkonů je uvedno na následujícím obrázku a je detailněji vysvětlováno ve zbytku článku.

![Konceptuální schéma úkonů v RPP](img/ukony-komplet.png)

## Zpřístupnění dat o úkonech v RPP

Data o úkonech jsou dostupná v datové sadě [Úkony](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F02925784ee989ac89fbb61d2e3b94c8e) a jejich struktura je zdokumentována v [dokumentaci](https://data.gov.cz/otevřené-formální-normy/registr-práv-a-povinností/úkony/2019-04-03/#json-přehled). Datová sada obsahuje pouze základní data o úkonech. Doplňující data jsou dostupná v [dalších datových sadách](https://data.gov.cz/datov%C3%A9-sady?kl%C3%AD%C4%8Dov%C3%A1%20slova=registr%20pr%C3%A1v%20a%20povinnost%C3%AD&str%C3%A1nka=0) z Registru práv a povinností. Na potřebné datové sady vždy upozorníme v příslušné části článku.

Každá datová sada z RPP je dostupná ve dvou podobách, tzv. distribucích. Ty jsou z věcného pohledu totožné. Dokonce jsou téměř totožné i z pohledu formátu (obě jsou ve formátu JSON). Liší se pouze datovým modelem. JSON distribuce jsou klasickými JSON soubory. [JSON-LD](https://json-ld.org/) distribuce jsou také JSON soubory, ale doplněné o tzv. [JSON-LD kontext](https://www.w3.org/TR/json-ld/#the-context), který mapuje JSON strukturu do datového modelu RDF s dodržením principů [otevřených propojených dat (angl. Linked Open Data)](https://lod-cloud.net/). S touto podobou lze pracovat přímo prostřednitvím veřejně dostupného [SPARQL endpointu](https://rpp-opendata.egon.gov.cz/odrpp/sparql). SPARQL endpoint je webová služba, která přijímá dotazy v dotazovacím jazyku [SPARQL](https://www.w3.org/TR/sparql11-overview/) a výsledky vrací buď v podobě tabulek serializovaných jako CSV, JSON nebo XML soubory nebo ve zvolené serializaci datového modelu [RDF](https://www.w3.org/RDF/).

## Práce s RPP prostřednictvím SPARQL endpointu

Svoji aplikaci můžete vystavět přímo nad [SPARQL endpointem](https://rpp-opendata.egon.gov.cz/odrpp/sparql). Nebudete tak muset budovat datový back-end. Na druhou stranu jste závislí na dostupnosti a výkonu veřejného SPARQL endpointu.

Prostřednictvím HTTPS GET/POST požadavků můžete na [SPARQL endpoint](https://rpp-opendata.egon.gov.cz/odrpp/sparql) zasílat SPARQL dotazy dle příkladů SPARQL dotazů uvedených níže. V případě SELECT dotazů jsou výsledkem tabulková data, která si můžete nechat vracet v různých formátech (CSV, XML, JSON, ...). Seznam možných formátů a způsob jak se na daný formát zeptat zjistíte pohledem na kód formulářového pole "Results Format" v [uživatelském rozhraní SPARQL endpointu](https://rpp-opendata.egon.gov.cz/odrpp/sparql).

Alternativou je zprovoznit si vlastní lokální RDF úložiště a využít jeho vlastní SPARQL endpoint pro dotazování. Lze využít např. [Apache Jena Fuseki](https://jena.apache.org/documentation/fuseki2/) nebo [GraphDB](http://graphdb.ontotext.com/). V Národním katalogu otevřených dat je potřeba získat odkazy ke stažení JSON-LD distribucí.

Pro spouštění SPARQL dotazů uvedených v tomto článku prostřednictvím prostřeí [Jupyter Notebooks](https://jupyter.org/) je nejprve nutno nastavit prostředí pro prácí s RDF a SPARQLem.

In [1]:
import  sys
from  gastrodon  import RemoteEndpoint,QName,ttl,URIRef,inline
import pandas as pd
pd.options.display.width=120
pd.options.display.max_colwidth=100

In [2]:
prefixes=inline("""
    @prefix l: <https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/> .
    @prefix a: <https://slovník.gov.cz/agendový/104/pojem/> .
""").graph

In [3]:
endpoint=RemoteEndpoint(
    "https://rpp-opendata.egon.gov.cz/odrpp/sparql"
    ,default_graph=""
    ,prefixes=prefixes
    ,base_uri=""
)

## Práce s RPP prostřednictvím JSON souborů

Jak jsme popsali výše, prostřednictvím [katalogizačních záznamů v NKOD](https://data.gov.cz/datov%C3%A9-sady?kl%C3%AD%C4%8Dov%C3%A1%20slova=registr%20pr%C3%A1v%20a%20povinnost%C3%AD) je možné stáhnout JSON distribuce datových sad z Registru práv a povinností a pro aplikaci využít tyto JSON soubory. Ty je samozřejmě nutno si uložit ve vhodném úložišti nabízejícím i vhodné dotazovací prostředí. Dokumentace každé z datových sad dokumentuje JSON strukturu a každý prvek JSON struktury odkazuje na odpovídající prvek konceptuálního schématu. Z konceptuálního pohledu je tak JSON reprezentace totožná s RDF reprezentací. Pouze v RDF reprezentaci je přirozeně zachycena grafová (síťová) struktura RPP, zatímco JSON struktury jsou z principu hierarchické a rozdělené do několika souborů. Práce s JSON podobou je tak v tomto ohledu s RPP trochu složitější.

## Příklady práce s daty

Ve zbytku článku si ukážeme možnosti práce s daty o úkonech v RPP. Jednotlivé podkapitoly se věnují jiné části dat o úkonech a popisují ji pomocí konceptuálního modelu, poskytují ukázku SPARQL dotazu či dotazů nad touto částí dat a specifikují odpovídající JSON strukturu pro ty, kteří nechtějí pracovat s RDF reprezentací a dotazovacím jazykem SPARQL. Specifikace JSON struktury je vyjádřena v notaci dotazů v dotazovacím jazyku [GraphQL](https://graphql.org/).

### Základní údaje o úkonech

O každém úkonu jsou v RPP vedeny následující základní údaje: identifikátor úkonu, název úkonu, popis úkonu, zda lze o výkon úkonu požádat elektronicky (vazba [Má elektronickou formu úkonu](https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/má-elektronickou-formu-úkonu)) a agenda, v rámci které je úkon vykonáván. Agenda označuje ucelenou oblast, ve které jeden nebo více OVM působí. Příklady agend jsou "Dohled nad finančním trhem", "Lidské tkáně a buňky" nebo "Veřejné zdravotní pojištění". Agenda je propojena s úkonem prostřednictvím vazby [Je vykonáván v agendě](https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/je-vykonáván-v-agendě).

Následující obrázek ukazuje konceptuální schéma základních údajů o úkonech.

![Základní údaje o úkonech](img/úkon-základ.png)

Popsané konceptuální schéma již jde nad rámec základních dat o úkonech. Zmiňuje také data o agendách (konkrétně jejich kód a název), které jsou dostupné v datové sadě [Agendy](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F9c73b802263c5e0ccf5542f10fbc35bb). To, jaká data tato datová sada obsahuje, je popsáno v [dokumentaci](https://data.gov.cz/otevřené-formální-normy/registr-práv-a-povinností/agendy/). Dále je pro práci potřeba datová sada [Číselník elektronických forem úkonů](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F420dc4fe7034cdb75bb0d1baf2c357d3) a její [dokumentace](https://data.gov.cz/otevřené-formální-normy/registr-práv-a-povinností/číselník-elektronických-forem-úkonů/).

#### SPARQL a RDF

Nejprve si pro ukázku vypíšeme seznam prvních 10 úkonů. K tomu slouží následující SPARQL dotaz.

In [4]:
úkony=endpoint.select("""PREFIX l: <https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/>
PREFIX a: <https://slovník.gov.cz/agendový/104/pojem/>

SELECT *
WHERE {
  ?ukon a l:úkon ;
    a:má-název-úkonu ?nazev .
}
LIMIT 10""")
úkony

ukon  \
0  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1147   
1  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1166   
2  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1167   
3  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1168   
4  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1170   
5  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1171   
6   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U417   
7   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U418   
8   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U419   
9   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U420   

                                                                                                 nazev  
0  Výzva orgánu územního plánování a stavebního úřadu vlastníkovi technické infrastruktury k poskyt...  
1        Uznávání odborné kvalifikace a jiné způsobilosti pro výkon povolání autorizovaného inspektora  
2                                                     Nařízení spojená se stavbou a stavebním pozemkem  
3  Zabezpečení součinnosti autorizovaného inspektora, znalce, vědeckého nebo jiného specializovanéh...  
4  Mimořádné postupy stavebního úřadu při vyhlašování stavu nebezpečí, nouzového stavu, stavu ohrož...  
5                                                                     Vstupy na pozemky nebo do staveb  
6                                         Podání podnětu k pořízení aktualizace zásad územního rozvoje  
7                                             Podání podnětu k pořízení územního plánu nebo jeho změny  
8                                          Podání podnětu k pořízení regulačního plánu nebo jeho změny  
9                                                                  Poskytnutí náhrady za změnu v území

Ve výsledku vidíme 10 úkonů. Pro každý úkon vidíme jeho IRI a název. Ve světě LOD je každá entita identifikována svým IRI. Toto IRI je zároveň dereferencovatelné, při jeho dereferencování dostaneme data o identifikované entitě.

Následující dotaz vrací základní údaje pro úkon U1073 "Zřízení nového válečného hrobu".

In [5]:
úkonyNázvy=endpoint.select("""PREFIX l: <https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/>
PREFIX a: <https://slovník.gov.cz/agendový/104/pojem/>

SELECT *
WHERE {
  VALUES (?ukon) {(<https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073>)}
  ?ukon a:má-název-úkonu ?nazev ;
    a:má-popis-úkonu ?popis ;
    l:má-identifikátor-úkonu ?id ;
    l:má-elektronickou-formu-úkonu ?elForma ;
    l:je-vykonáván-v-agendě ?agenda .
}""")
úkonyNázvy

ukon                           nazev  \
0  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  Zřízení nového válečného hrobu   

                                                                                                 popis     id  \
0  Na základě písemné žádosti zřizovatele nového válečného hrobu a po písemném souhlasu vlastníka n...  U1073   

                                                                     elForma  \
0  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/elektronická-forma-úkonu/ANO   

                                                     agenda  
0  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/agenda/A549

Elektronická forma úkonu a agenda, ve které je úkon vykonáván, jsou entitami, pro které obsahuje výsledná tabulka jejich IRI. Toto IRI je, opět, dle LOD principů dereferencovatelné. Dotaz lze ale upravit tak, aby nevracel IRI těchto entit, ale jejich názvy. K agendám navíc můžeme přidat i jejich identifikační kód. IRI entit potřebujeme pro snadné propojování s dalšími entitami. Pokud ale elektronickou formu ani agendu nechceme propojovat dále na další související entity, IRI nepotřebujeme. Viz následující dotaz.

In [6]:
úkonZákladníDetail=endpoint.select("""PREFIX l: <https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/>
PREFIX a: <https://slovník.gov.cz/agendový/104/pojem/>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>

SELECT ?nazev ?popis ?id ?elFormaNazev ?agendaNazev ?agendaKod
WHERE {
  VALUES (?ukon) {(<https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073>)}
  ?ukon a:má-název-úkonu ?nazev ;
    a:má-popis-úkonu ?popis ;
    l:má-identifikátor-úkonu ?id ;
    l:má-elektronickou-formu-úkonu ?elForma ;
    l:je-vykonáván-v-agendě ?agenda .
    
  ?elForma skos:prefLabel ?elFormaNazev .
  
  ?agenda l:má-název-agendy ?agendaNazev ;
    l:má-kód-agendy ?agendaKod .
}""")
úkonZákladníDetail

nazev  \
0  Zřízení nového válečného hrobu   

                                                                                                 popis     id  \
0  Na základě písemné žádosti zřizovatele nového válečného hrobu a po písemném souhlasu vlastníka n...  U1073   

  elFormaNazev           agendaNazev agendaKod  
0          Ano  Péče o válečné hroby      A549

#### JSON

V případě práce s JSON soubory je možné tyto základní údaje o úkonech získat z JSON distribuce datové sady [Úkony](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F02925784ee989ac89fbb61d2e3b94c8e). V JSON souboru jsou výše popsané údaje kódovány v následující JSON struktuře (viz [dokumentace](https://data.gov.cz/otevřené-formální-normy/registr-práv-a-povinností/úkony/) pro jejich detailní popis):

```
{
  položky {
    identifikátor,
    název,
    popis,
    agenda,
    elektronická-forma
  }
}
```

Název agendy (příp. další atributy agend) lze získat z JSON distribuce datové sady [Agenda](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F9c73b802263c5e0ccf5542f10fbc35bb). V JSON souboru je identifikátor agendy odkazovaný z úkonu, kód a název agendy kódován v následující JSON struktuře (viz [dokumentace](https://data.gov.cz/otevřené-formální-normy/registr-práv-a-povinností/úkony/) pro jejich detailní popis):

```
{
  položky {
    id,
    kód
    název
  }
}
```

### Poskytování úkonů

OVM poskutuje úkony. To, že daný OVM poskytuje úkon, je v RPP zaznamenáno buď přímo nebo je zde uvedeno, že úkon je poskytován kategorií OVM. Úkon pak vykonává každý OVM, který do kategorie patří. Konceptuální schéma poskytování úkonů je znázorněno na následujícím obrázku.

![Konceptuální schéma poskytování úkonů](img/poskytování-úkonů-01.png)

Vazba mezi úkonem a OVM, resp. kategorií OVM, není jednoduchou binární vazbou. Do vazby totiž vstupuje ještě další prvek, o kterém si řekneme až později. Pro nyní není podstatný. Z důvodu tohoto dalšího prvku je vazba ternární a je proto reprezentována samostatnou třídou [Poskytování úkonu](https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/poskytování-úkonu). Třída [Úkon](https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/úkon) je pak propojena s třídou [Poskytování úkonu](https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/poskytování-úkonu) prostřednictvím vazby [Je poskytován](https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/poskytování-úkonu). S třídami [Orgán veřejné moci](https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/orgán-veřejné-moci) a [Kategorie orgánů veřejné moci](https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/kategorie-orgánů-veřejné-moci) je propojena prostřednictvím vazby [Má poskytovatele](https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/má-poskytovatele).

V případě, že úkon poskytuje celá kategorie OVM, je potřeba také získat OVM, které do kategorie patří. Vazba mezi kategorií a OVM není jednoduchou binární vazbou, neboť vztah mezi OVM a kategorií má svoje atributy. Konceptuální model zařazení OVM do kategorií je znározněn na následujícím obrázku.

![Zařazení OVM do kategorií](img/ovm-a-kategorie.png)

Pro práci s daty o OVM a jejich zařazení do kategorií je potřebná datová sada [Orgány veřejné moci](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F44a9d6abacd4d0e83a0694e74d028f51) a její [dokumentace](https://data.gov.cz/otevřené-formální-normy/registr-práv-a-povinností/orgány-veřejné-moci/). Pro práci s daty o kategoriích OVM je potřebná datová sada [Kategorie orgánů veřejné moci](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F4cdc3767e173806c2a7bb345459731a5) a její [dokumentace](https://data.gov.cz/otevřené-formální-normy/registr-práv-a-povinností/kategorie-orgánů-veřejné-moci/).

#### SPARQL a RDF

Následující dotaz vrací OVM či jejich kategorie, které poskytují úkon U1073 "Zřízení nového válečného hrobu".

In [7]:
poskytujícíOVMaKategorie=endpoint.select("""PREFIX l: <https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/>
PREFIX a: <https://slovník.gov.cz/agendový/104/pojem/>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>

SELECT DISTINCT ?ukon ?id ?ovmNeboKategorie ?ovmIc ?ovmNazev ?kategorieId ?kategorieNazev
WHERE {
  VALUES (?ukon) {(<https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073>)}
  
  ?ukon l:má-identifikátor-úkonu ?id ;
    l:je-poskytován ?poskytovani .

  ?poskytovani l:má-poskytovatele ?ovmNeboKategorie .

  {
    ?ovmNeboKategorie l:má-identifikační-číslo-osoby-orgánu-veřejné-moci ?ovmIc ;
      l:má-název-orgánu-veřejné-moci ?ovmNazev .
  } UNION {
    ?ovmNeboKategorie l:má-identifikátor-kategorie ?kategorieId ;
      l:má-název-kategorie ?kategorieNazev .
  }
}""")
poskytujícíOVMaKategorie

ukon     id  \
0  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
1  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
2  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
3  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   

                                                           ovmNeboKategorie     ovmIc             ovmNazev  \
0  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/orgán-veřejné-moci/60162694  60162694  Ministerstvo obrany   
1           https://rpp-opendata.egon.gov.cz/odrpp/zdroj/kategorie-ovm/KO11      None                 None   
2           https://rpp-opendata.egon.gov.cz/odrpp/zdroj/kategorie-ovm/KO12      None                 None   
3          https://rpp-opendata.egon.gov.cz/odrpp/zdroj/kategorie-ovm/KO241      None                 None   

  kategorieId                kategorieNazev  
0        None                          None  
1        KO11  Obce s rozšířenou působností  
2        KO12          Kraje a hl. m. Praha  
3       KO241    Městské části Prahy 1 – 22

Následující dotaz vrací OVM zařazené do kategorie KO12 "Kraje a hl. m. Praha".

In [8]:
ovmVKategorii=endpoint.select("""PREFIX l: <https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/>
PREFIX a: <https://slovník.gov.cz/agendový/104/pojem/>

SELECT DISTINCT ?kategorie ?kategorieId ?kategorieNazev ?ovm ?ovmIc ?ovmNazev
WHERE {
  VALUES (?kategorie) {(<https://rpp-opendata.egon.gov.cz/odrpp/zdroj/kategorie-ovm/KO12>)}
  
  ?zarazeni a:zařazuje-do-kategorie ?kategorie ;
    a:má-zařazený-subjekt ?ovm .

  OPTIONAL {
    ?zarazeni a:má-datum-zařazení-do-kategorie ?od .
  }

  OPTIONAL {
    ?zarazeni a:má-datum-vyřazení-z-kategorie ?do .
  }

  ?ovm a l:orgán-veřejné-moci ;
    l:má-identifikační-číslo-osoby-orgánu-veřejné-moci ?ovmIc ;
    l:má-název-orgánu-veřejné-moci ?ovmNazev .

  ?kategorie l:má-identifikátor-kategorie ?kategorieId ;
    l:má-název-kategorie ?kategorieNazev .
}""")
ovmVKategorii

kategorie kategorieId        kategorieNazev  \
0   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/kategorie-ovm/KO12        KO12  Kraje a hl. m. Praha   
1   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/kategorie-ovm/KO12        KO12  Kraje a hl. m. Praha   
2   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/kategorie-ovm/KO12        KO12  Kraje a hl. m. Praha   
3   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/kategorie-ovm/KO12        KO12  Kraje a hl. m. Praha   
4   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/kategorie-ovm/KO12        KO12  Kraje a hl. m. Praha   
5   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/kategorie-ovm/KO12        KO12  Kraje a hl. m. Praha   
6   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/kategorie-ovm/KO12        KO12  Kraje a hl. m. Praha   
7   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/kategorie-ovm/KO12        KO12  Kraje a hl. m. Praha   
8   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/kategorie-ovm/KO12        KO12  Kraje a hl. m. Praha   
9   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/kategorie-ovm/KO12        KO12  Kraje a hl. m. Praha   
10  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/kategorie-ovm/KO12        KO12  Kraje a hl. m. Praha   
11  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/kategorie-ovm/KO12        KO12  Kraje a hl. m. Praha   
12  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/kategorie-ovm/KO12        KO12  Kraje a hl. m. Praha   
13  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/kategorie-ovm/KO12        KO12  Kraje a hl. m. Praha   

                                                                         ovm     ovmIc              ovmNazev  
0   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/orgán-veřejné-moci/00064581     64581    HLAVNÍ MĚSTO PRAHA  
1   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/orgán-veřejné-moci/70890650  70890650        Jihočeský kraj  
2   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/orgán-veřejné-moci/70889546  70889546  Královéhradecký kraj  
3   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/orgán-veřejné-moci/70890692  70890692  Moravskoslezský kraj  
4   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/orgán-veřejné-moci/60609460  60609460        Olomoucký kraj  
5   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/orgán-veřejné-moci/70890366  70890366         Plzeňský kraj  
6   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/orgán-veřejné-moci/70890749  70890749         Kraj Vysočina  
7   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/orgán-veřejné-moci/70888337  70888337     Jihomoravský kraj  
8   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/orgán-veřejné-moci/70891320  70891320          Zlínský kraj  
9   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/orgán-veřejné-moci/70892822  70892822       Pardubický kraj  
10  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/orgán-veřejné-moci/70891168  70891168      Karlovarský kraj  
11  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/orgán-veřejné-moci/70891095  70891095      Středočeský kraj  
12  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/orgán-veřejné-moci/70891508  70891508        Liberecký kraj  
13  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/orgán-veřejné-moci/70892156  70892156          Ústecký kraj

#### JSON

V případě práce s JSON soubory lze OVM a kategorie OVM poskytující úkony získat z JSON distribuce datové sady [Úkony](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F02925784ee989ac89fbb61d2e3b94c8e), kde jsou kódovány v následující JSON struktuře (viz [dokumentace](https://data.gov.cz/otevřené-formální-normy/registr-práv-a-povinností/úkony/) pro jejich detailní popis):

```
{
  položky {
    poskytování-úkonu {
      poskytovatel
    }
  }
}
```

kde hodnotou prvku ```poskytovatel``` je buď id OVM nebo id kategorie OVM.

Pro získání údajů o OVM je nutno přistoupit do JSON distribuce datové sady [Orgány veřejné moci](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F44a9d6abacd4d0e83a0694e74d028f51), kde jsou identifikátor OVM referencovaný z úkonu, název OVM a IČ osoby OVM kódovány v následující JSON struktuře (viz [dokumentace](https://data.gov.cz/otevřené-formální-normy/registr-práv-a-povinností/orgány-veřejné-moci/) pro jejich detailní popis):

```
{
  položky {
    id,
    název,
    ičo
  }
}
```

Pro získání údajů o kategorii OVM je nutno přistoupit do JSON distribuce datové sady [Kategorie orgánů veřejné moci](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F4cdc3767e173806c2a7bb345459731a5), kde jsou identifikátor kategorie OVM referencovaný z úkonu, název OVM a jeho v RPP evidovaný identifikátor kódovány v následující JSON struktuře (viz [dokumentace](
https://data.gov.cz/otevřené-formální-normy/registr-práv-a-povinností/kategorie-orgánů-veřejné-moci/) pro jejich detailní popis):

```
{
  položky {
    id,
    název,
    identifikátor
  }
}
```

U kategorií OVM je nutno odlišit ```id``` a ```identifikátor```. První je identifikátor používaný k referencování mezi datovými sadami. Druhý je identifikátor tak, jak je uveden v RPP.

Dále pak pro získání kategorií, do kterých OVM spadá je nutno přistoupit do JSON distribuce datové sady [Orgány veřejné moci](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F44a9d6abacd4d0e83a0694e74d028f51), kde je uveden seznam identifikátu kategorií, do kterých OVM patří kódován v následující JSON struktuře (viz [dokumentace](https://data.gov.cz/otevřené-formální-normy/registr-práv-a-povinností/orgány-veřejné-moci/) pro jejich detailní popis):

```
{
  položky {
    seznam-kategorií {
      kategorie,
      zařazení-od,
      zařazení-do
    }
  }
}
```

Prvek ```kategorie``` je referencí na kategorii v datové sadě datové sady [Kategorie orgánů veřejné moci](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F4cdc3767e173806c2a7bb345459731a5).

### Vykonávání úkonů

Poskytování úkonu vyplývá z legislativy, která říká, že daný OVM nebo OVM dané kategorie mají úkon poskytovat. Ještě to ale neznamená, že úkon skutečně poskytují. Může například dojít k tomu, že určitý OVM ve skutečnosti úkon, který má poskytovat, přenese na jiný OVM. Skutečnost je pak v RPP zaznamenána v podobě výkonu úkonů. V tomto případě jsou zaznamenávány výkony úkonů na detail jednotlivých OVM, nikoliv na úroveň kategorií. Tj. jestliže je v RPP zaznamenáno, že daná kategorie poskytuje úkon, pak je v RPP také zaznamenáno, že každý jednotlivý OVM z této kategorie úkon vykonává, pokud výkon úkonu nepřenesl tento OVM na jiný OVM. Konceptuální schéma poskytování úkonů je znázorněno na následujícím obrázku.

![Konceptuální schéma výkonu úkonů](img/vykonávání-úkonů-01.png)

Vazba mezi úkonem a OVM opět není jednoduchou binární vazbou. Do vazby opět vstupují další prvek, které nyní pro jednoduchost pomineme. Výkon úkonu probíhá v rámci tzv. působnosti OVM v agendě. OVM je pro výkon agendy registrovaný. Registrace OVM pro výkon agendy v RPP znamená, že OVM se oficiálně přihlašuje k tomu, že agendu skutečně vykonává. Registrace je na konceptuální úrovni reprezentována třídou [Registrace pro výkon agendy](https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/registrace-pro-výkon-agendy) a je propojena s třídou [Orgán veřejné moci ](https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/orgán-veřejné-moci) prostřednictvím vazby [Má registrovaný subjekt](https://slovník.gov.cz/agendový/104/pojem/má-registrovaný-subjekt). Registrace je navázaná i na samostatnou agendu, ale tato vazba teď není podstatná. Registrace znamená, že OVM vykonává i některé úkony vykonávané v rámci agendy (viz dotazy výše). Registrací vzniká působnost v agendě reprezentovaná třídou [Působnost v agendě](https://slovník.gov.cz/agendový/104/pojem/působnost-v-agendě). Na registraci je navázaná vazbou [Vznikla v rámci registrace](https://slovník.gov.cz/agendový/104/pojem/vznikla-v-rámci-registrace). V rámci působnosti pak již registrovaný OVM vykonává jednotlivé úkony. Výkony úkonu jsou reprezentovány třídou [Výkon úkonu](https://slovník.gov.cz/agendový/104/pojem/výkon-úkonu) a na úkony jsou navázány vazbou [Má vykonávaný úkon](https://slovník.gov.cz/agendový/104/pojem/má-vykonávaný-úkon).

Tuto cestu mezi úkony a OVM, které je vykonávají, je nutno zahrnout do dotazů nad daty. Ještě poznamenejme, že úkon může vykonávat také tzv. soukromoprávní uživatel údajů, ale to teď pro jednoduchost vynechme a uvažujme pro výkon úkonů pouze OVM.

Data o výkonech úkonů v rámci působnosti v agendách jsou dostupná v datové sadě [Působnost v agendách](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F7ee91a644bf6da16a9f5c4f337163c0f). Pro práci s datovou sadou využijte [dokumentaci](https://data.gov.cz/otevřené-formální-normy/registr-práv-a-povinností/působnost-v-agendách/). Je nutno upozornit, že se jedná o poměrně velkou datovou sadu čítající 634730 působností, přičemž 1 působnost je poměrně složitým objektem (viz [dokumentace](https://data.gov.cz/otevřené-formální-normy/registr-práv-a-povinností/působnost-v-agendách/)).

#### SPARQL a RDF

Následující SPARQL dotaz vrací OVM, které vykonávají úkon U1073 "Zřízení nového válečného hrobu".

In [9]:
vykonávajícíOVMaKategorie=endpoint.select("""PREFIX l: <https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/>
PREFIX a: <https://slovník.gov.cz/agendový/104/pojem/>

SELECT DISTINCT ?ukon ?id ?ovm ?ovmIc ?ovmNazev
WHERE {
  VALUES (?ukon) {(<https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073>)}

  ?pusobnost a:vznikla-v-rámci-registrace/a:má-registrovaný-subjekt ?ovm ;
    a:vykonává-úkon-v-rámci-působnosti/a:má-vykonávaný-úkon ?ukon .

  ?ovm a l:orgán-veřejné-moci ;
    l:má-identifikační-číslo-osoby-orgánu-veřejné-moci ?ovmIc ;
    l:má-název-orgánu-veřejné-moci ?ovmNazev .

  ?ukon l:má-identifikátor-úkonu ?id .

}""")
vykonávajícíOVMaKategorie

ukon     id  \
0    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
1    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
2    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
3    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
4    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
5    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
6    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
7    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
8    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
9    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
10   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
11   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
12   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
13   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
14   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
15   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
16   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
17   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
18   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
19   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
20   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
21   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
22   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
23   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
24   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
25   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
26   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
27   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
28   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
29   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
..                                                       ...    ...   
212  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
213  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
214  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
215  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
216  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
217  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
218  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
219  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
220  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
221  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
222  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
223  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
224  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
225  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
226  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
227  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
228  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
229  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
230  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
231  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
232  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
233  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
234  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
235  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
236  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
237  http

#### JSON

V případě práce s JSON soubory lze OVM vykonávající úkony získat z JSON distribuce datové sady [Působnost v agendách](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F7ee91a644bf6da16a9f5c4f337163c0f) (viz [dokumentace](https://data.gov.cz/otevřené-formální-normy/registr-práv-a-povinností/působnost-v-agendách/) pro detailní popis struktury). Je nutno upozornit, že se jedná o poměrně velkou datovou sadu čítající 634730 působností, přičemž 1 působnost je poměrně složitým objektem. Prvek ```položky``` v JSON souboru je seznamem (polem) jednotlivých působností v agendách. Každá působnost je navázána na OVM a na úkony, které OVM v rámci této působnosti vykonává. OVM a úkony jsou kódovány v následující JSON struktuře:

```
{
  položky {
    registrace {
      ovm
    },
    úkony {
      úkon
    }
  }
}
```

kde hodnotou prvku ```ovm``` je reference na id OVM do datové sady [Orgány veřejné moci](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F44a9d6abacd4d0e83a0694e74d028f51) a hodnotou prvku ```úkon``` je reference na id úkonu do datové sady [Úkony](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F02925784ee989ac89fbb61d2e3b94c8e).

### Pracoviště, kde jsou úkony vykonávány

Pokud OVM vykonává úkon na konkrétních pracovištích, pak jsou také evidována v RPP. Souvislost mezi výkonem úkonu a pracovištěm vykonávajícího OVM je znázorněna na následujícím konceptuálním schématu.

![Konceptuální schéma výkonu úkonů na pracovištích](img/vykonávání-úkonů-02.png)

Pracoviště jsou reprezentovánav koncetuálním modelu třídou [Pracoviště](https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/pracoviště). Pracoviště jsou propojena s výkonem úkonu vazbou [Vykonává úkon na pracovišti](https://slovník.gov.cz/agendový/104/pojem/vykonává-úkon-na-pracovišti). Pracoviště je prostřednictvím vazby [Má adresu pracoviště](https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/má-adresu-pracoviště) propojeno na svoji adresu. Ta může být uvedena buď v podobě adresního místa RÚIAN nebo jako textová adresa, pokud se jedná o adresu, která není evidována v RÚIAN. Dále je pracoviště propojeno prostřednictvím vazby [Má stát pracoviště](https://slovník.gov.cz/agendový/104/pojem/má-stát-pracoviště) na stát, ve kterém se pracoviště nachází. Stát je uveden pouze v případě zahraničních pracovišť (např. velvyslanectví). Orgán veřejné moci je asociován se seznamem svých pracovišť vazbou [Má pracoviště vykonávající úkon](https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/má-pracoviště-vykonávající-úkon).

Vazby pracovišť na výkony úkonů jsou dostupné v datové sadě [Působnost v agendách](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F7ee91a644bf6da16a9f5c4f337163c0f). Pro práci s datovou sadou využijte [dokumentaci](https://data.gov.cz/otevřené-formální-normy/registr-práv-a-povinností/působnost-v-agendách/). Data o pracovištích jsou dostupná v datové sadě [Orgány veřejné moci](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F44a9d6abacd4d0e83a0694e74d028f51) a použijte [dokumentaci](https://data.gov.cz/otevřené-formální-normy/registr-práv-a-povinností/orgány-veřejné-moci/).

#### SPARQL a RDF

Následující dotaz vrací pracoviště, na kterých je vykonáván úkon U1073 "Zřízení nového válečného hrobu".

In [10]:
vykonávajícíPracoviště=endpoint.select("""PREFIX l: <https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/>
PREFIX a: <https://slovník.gov.cz/agendový/104/pojem/>

SELECT DISTINCT ?ukon ?id ?ovm ?ovmIc ?adresniMisto ?adresaNonRuian ?stat
WHERE {
  VALUES (?ukon) {(<https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073>)}

  ?pusobnost a:vznikla-v-rámci-registrace/a:má-registrovaný-subjekt ?ovm ;
    a:vykonává-úkon-v-rámci-působnosti ?vykon .

  ?vykon a:má-vykonávaný-úkon ?ukon ;
    a:vykonává-úkon-na-pracovišti ?pracoviste .

  OPTIONAL {
    ?pracoviste l:má-adresu-pracoviště ?adresniMisto .
  }

  OPTIONAL {
    ?pracoviste a:má-adresu-pracoviště-neevidovanou-v-rúian ?adresaNonRuian .
  }

  OPTIONAL {
    ?pracoviste a:má-stát-pracoviště ?stat .
  }

  ?ovm a l:orgán-veřejné-moci ;
    l:má-pracoviště-vykonávající-úkon ?pracoviste ;
    l:má-identifikační-číslo-osoby-orgánu-veřejné-moci ?ovmIc .

  ?ukon l:má-identifikátor-úkonu ?id .

}""")
vykonávajícíPracoviště

ukon     id  \
0    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
1    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
2    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
3    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
4    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
5    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
6    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
7    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
8    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
9    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
10   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
11   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
12   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
13   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
14   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
15   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
16   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
17   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
18   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
19   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
20   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
21   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
22   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
23   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
24   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
25   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
26   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
27   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
28   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
29   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
..                                                       ...    ...   
201  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
202  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
203  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
204  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
205  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
206  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
207  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
208  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
209  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
210  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
211  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
212  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
213  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
214  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
215  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
216  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
217  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
218  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
219  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
220  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
221  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
222  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
223  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
224  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
225  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
226  http

Z výsledku můžeme vyčíst, že adres, na kterých je úkon vykonáván je méně, než OVM, které úkon vykonávají. To je způsobeno určitou děravostí RPP, která je vidět např. na Hlavním městě Praze (IČ 00064581). Hlavní město Praha má evidována celkem [4 pracoviště](https://rpp-opendata.egon.gov.cz/odrpp/zdroj/orgán-veřejné-moci/00064581). Pokud se ale podíváme na [působnost Hlavního města Prahy, v rámci které vykonává úkon U1073](https://rpp-opendata.egon.gov.cz/odrpp/zdroj/působnost/A549/00064581), zjistíme, že pracoviště není uvedeno.

#### JSON

V případě práce s JSON soubory lze pracoviště získat z JSON distribuce datové sady [Působnost v agendách](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F7ee91a644bf6da16a9f5c4f337163c0f) (viz [dokumentace](https://data.gov.cz/otevřené-formální-normy/registr-práv-a-povinností/působnost-v-agendách/) pro detailní popis struktury). Pracoviště jsou uvedeny vedle vazby působnosti na úkon v následující JSON struktuře:

```
{
  položky {
    úkony {
      pracoviště
    }
  }
}
```

kde hodnotou prvku ```pracoviště``` je reference na id pracoviště do datové sady [Orgány veřejné moci](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F44a9d6abacd4d0e83a0694e74d028f51), kde je uvedeno v následující struktuře:

```
{
  položky {
    pracoviště-ovm {
      id,
      adresa-pr,
      adresa-pr-txt,
      stat-pr
    }
  }
}
```

kde referencované id pracoviště je uvedeno v prvku ```id```.

### GPS souřadnice a další údaje adresních míst pracovišť

Z výsledku předchozího dotazu také můžeme vyčíst, že adresy pracovišť jsou uvedeny v podobě referencí na adresní místa RÚIAN. V otevřených datech z RPP ale údaje o těchto adresách dostupné nejsou a je nutno využít otevřená data z RÚIAN.

Níže je popsána pouze práce s RÚIAN prostřednictvím dotazovacího jazyka SPARQL. ČÚZK poskytuje RÚIAN svými vlastními prostředky buď prostřednictvím [datových sad](https://data.gov.cz/datov%C3%A9-sady?kl%C3%AD%C4%8Dov%C3%A1%20slova=R%C3%9AIAN) v NKOD nebo prostřednictvím [Veřejného dálkového přístupu k RÚIAN](https://vdp.cuzk.cz/).

V oblasti LOD ČÚZK stanovil tvar IRI jednotlivých RÚIAN objektů, včetně adresních míst, ale data v podobě LOD ještě neposkytuje. Nicméně, neoficiálně jsou data RÚIANu v LOD podobě dostupná prostřednictvím [SPARQL endpointu iniciativy OpenData.cz](https://linked.cuzk.cz.opendata.cz/sparql).

In [11]:
endpointRUIAN=RemoteEndpoint(
    "https://linked.cuzk.cz.opendata.cz/sparql"
)

Zde můžeme zjistit např. GPS souřadnice daného adresního místa. Jsou zde totiž nad rámec dat z RÚIAN data rovnou geokódovaná do GPS souřadnic. Následující dotaz vrátí GPS souřadnice adresního místa s kódem 12917931.

In [12]:
adresníMístoGPS=endpointRUIAN.select("""
PREFIX geo: <http://www.opengis.net/ont/geosparql#>
PREFIX s: <http://schema.org/>
SELECT ?adresniMisto ?lat ?lon
WHERE {
  VALUES ?adresniMisto {<https://linked.cuzk.cz/resource/ruian/adresni-misto/22305190>}
  
  ?adresniMisto geo:hasGeometry/s:geo ?gps .
  
  ?gps s:latitude ?lat ;
    s:longitude ?lon .
}
""")
adresníMístoGPS

adresniMisto        lat        lon
0  https://linked.cuzk.cz/resource/ruian/adresni-misto/22305190  50.098474  14.422026

Kromě GPS souřadnice lze k adresnímu místu získat i kompletní textovou informaci o adrese pomocí následujícího SPARQL dotazu. Ten reflektuje strukturu adres tak, jak je vedena v RÚIAN. Adresní místo má svoje číslo domovní a číslo orientační. Může mít ulici, ale ne nutně. Malé obce ulice nemají. Na adresním místě je pak stavení objekt, který je propojen na příslušnou část obce a ta na obec.

In [13]:
adresníMístoDetail=endpointRUIAN.select("""
PREFIX ruian: <https://linked.cuzk.cz/ontology/ruian/>
PREFIX s: <http://schema.org/>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>

SELECT *
WHERE {
  VALUES ?adresniMisto {<https://linked.cuzk.cz/resource/ruian/adresni-misto/22305190>}
  
  ?adresniMisto skos:notation ?adresniMistoKod .
  
  OPTIONAL {
    ?adresniMisto ruian:čísloOrientační ?cisloOrientacni .
  }

  OPTIONAL {    
    ?adresniMisto ruian:čísloDomovní ?cisloDomovni .
  }

  OPTIONAL {
    ?adresniMisto ruian:ulice ?ulice .
    ?ulice s:name ?uliceNazev .
  }
  
  ?adresniMisto ruian:stavebníObjekt/ruian:částObce ?castObce .
  
  ?castObce s:name ?castObceNazev ;
    ruian:obec ?obec .
    
  ?obec s:name ?obecNazev .

}
""")
adresníMístoDetail

adresniMisto  adresniMistoKod  cisloOrientacni  cisloDomovni  \
0  https://linked.cuzk.cz/resource/ruian/adresni-misto/22305190         22305190                3           936   

                                                ulice  uliceNazev  \
0  https://linked.cuzk.cz/resource/ruian/ulice/461521  Nad štolou   

                                                 castObce castObceNazev  \
0  https://linked.cuzk.cz/resource/ruian/cast-obce/490067    Holešovice   

                                                obec obecNazev  
0  https://linked.cuzk.cz/resource/ruian/obec/554782     Praha

### Subjekty využívající úkon

Pro každý úkon je uvedeno, jaké typy subjektů mohou úkon využívat. Přesněji to znamená vymezení subjektu, který může podat žádost o výkon úkonu. Jsou rozlišovány pouze 3 typy subjektů - fyzická osoba (FO), podnikající fyzická osoba (PFO) a právnická osoba (PO). Vazbu mezi úkonem a typem subjektů, které úkon využívají znázorňuje následující konceptuální schéma.

![Subjekty využívající úkon](img/typ-subjektu-k-úkonu.png)

Pro práci s typy subjektů je ještě potřeba datová sada [Číselník typů subjektů](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F281575084358cb8be628b7cc66020c2c) a její [dokumentace](https://data.gov.cz/otevřené-formální-normy/registr-práv-a-povinností/číselník-typů-subjektů/).

#### SPARQL

Následující dotaz vrací, jaké typy subjektů mohou využívat úkon U1073 "Zřízení nového válečného hrobu".

In [14]:
typySubjektů=endpoint.select("""PREFIX l: <https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/>
PREFIX a: <https://slovník.gov.cz/agendový/104/pojem/>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>

SELECT ?ukon ?id ?typSubjektuKod
WHERE {
  VALUES (?ukon) {(<https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073>)}

  ?ukon a:je-využíván-subjektem ?typSubjektu ;
    l:má-identifikátor-úkonu ?id .
  
  ?typSubjektu skos:notation ?typSubjektuKod .

}""")
typySubjektů

ukon     id typSubjektuKod
0  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073             FO
1  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073             PO

#### JSON

V případě práce s JSON soubory je možné typy subjektů navázané k úkonu získat z JSON distribuce datové sady [Úkony](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F02925784ee989ac89fbb61d2e3b94c8e), kde je vazba mezi úkonem a typy subjektů kódována v následující JSON struktuře (viz [dokumentace](https://data.gov.cz/otevřené-formální-normy/registr-práv-a-povinností/úkony/) pro jejich detailní popis):

```
{
  položky {
    subjekty-využívající-úkon
  }
}
```

Prvek ```subjekty-využívající-úkon``` je seznamem referencí na identifikátory typů subjektů uvedených v datové sadě [Číselník typů subjektů](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F281575084358cb8be628b7cc66020c2c).

### Místní příslušnost úkonu

Místo výkonu úkonu může být pro žadatele omezeno. Pokud je omezeno, potom žadatel může žádat o výkon úkonu pouze na pracovištích odpovídajících tomuto místu.

Omezení místa je v konceptuálním schématu RPP reprezentováno třídou [Místní příslušnost](https://slovník.gov.cz/agendový/104/pojem/místní-příslušnost). K místní příslušnosti jsou k dispozici následující údaje - vazba [Je místní příslušností pro typ subjektu](https://slovník.gov.cz/agendový/104/pojem/je-místní-příslušností-pro-typ-subjektu) odkazující na typ subjektu, kterého se omezení týká, vazba [Má poskytovatele v místní příslušnosti](https://slovník.gov.cz/agendový/104/pojem/má-poskytovatele-v-místní-příslušnosti) odkazující na OVM nebo kategorii OVM, která v dané místní příslušnosti úkon vykonává a vazba [Má typ místní příslušnosti](https://slovník.gov.cz/agendový/104/pojem/má-typ-místní-příslušnosti) vymezující, jak je místo výkonu omezeno. Konceptuální schéma místních přísluností úkonu ukazuje následující obrázek.

![Místní příslušnosti úkonů](img/místní-příslušnost-úkonu.png)

Pro práci s typy místních příslušností je ještě potřeba datová sada [Číselník typů místních příslušností k subjektům](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2Fcf79f7477c758122a8d99891e66a9800) a její [dokumentace](https://data.gov.cz/otevřené-formální-normy/registr-práv-a-povinností/číselník-typů-místních-příslušností-k-subjektům/).

#### SPARQL a RDF

Existující typy místních příslušností vrací následující dotaz.

In [15]:
typyMístníPříslušnosti=endpoint.select("""PREFIX skos: <http://www.w3.org/2004/02/skos/core#>

SELECT ?kod ?nazev
WHERE {
  ?typMistniPrislunosti a <https://slovník.gov.cz/agendový/104/pojem/typ-místní-příslušnosti-k-subjektu> ;
    skos:notation ?kod ;
    skos:prefLabel ?nazev .
}""")
typyMístníPříslušnosti

kod                                    nazev
0  M_CINNOST             Dle místa činnosti účastníka
1  M_NEMOVIT            Dle místa dotčené nemovitosti
2    M_POBYT  Dle místa trvalého pobytu fyzické osoby
3   M_PODNIK        Dle místa podnikání fyzické osoby
4       NENI                                     Není
5   SIDLO_PO                Dle sídla právnické osoby

Další dotaz vrací místní příslušnosti k úkonu U1073 "Zřízení nového válečného hrobu".

In [16]:
místníPříslušnostiKÚkonu=endpoint.select("""PREFIX l: <https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/>
PREFIX a: <https://slovník.gov.cz/agendový/104/pojem/>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>

SELECT ?ukon ?id ?typSubjektuKod ?typMistniPrislusnostiKod ?ovmNeboKategorie ?ovmIc ?ovmNazev ?kategorieId ?kategorieNazev
WHERE {
  VALUES (?ukon) {(<https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073>)}

  ?ukon a:má-místní-příslušnost ?prislusnost ;
    l:má-identifikátor-úkonu ?id .
  
  ?prislusnost a:je-místní-příslušností-pro-typ-subjektu ?typSubjektu ;
    a:má-poskytovatele-v-místní-příslušnosti ?ovmNeboKategorie ;
    a:má-typ-místní-příslušnosti ?typMistniPrislusnosti .

  ?typSubjektu skos:notation ?typSubjektuKod .

  ?typMistniPrislusnosti skos:notation ?typMistniPrislusnostiKod .

  {
    ?ovmNeboKategorie l:má-identifikační-číslo-osoby-orgánu-veřejné-moci ?ovmIc ;
      l:má-název-orgánu-veřejné-moci ?ovmNazev .
  } UNION {
    ?ovmNeboKategorie l:má-identifikátor-kategorie ?kategorieId ;
      l:má-název-kategorie ?kategorieNazev .
  }

}""")
místníPříslušnostiKÚkonu

ukon     id typSubjektuKod typMistniPrislusnostiKod  \
0  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073             FO                     NENI   
1  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073             PO                     NENI   
2  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073             FO                M_NEMOVIT   
3  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073             FO                M_NEMOVIT   
4  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073             FO                M_NEMOVIT   
5  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073             PO                M_NEMOVIT   
6  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073             PO                M_NEMOVIT   
7  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073             PO                M_NEMOVIT   

                                                           ovmNeboKategorie     ovmIc             ovmNazev  \
0  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/orgán-veřejné-moci/60162694  60162694  Ministerstvo obrany   
1  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/orgán-veřejné-moci/60162694  60162694  Ministerstvo obrany   
2           https://rpp-opendata.egon.gov.cz/odrpp/zdroj/kategorie-ovm/KO11      None                 None   
3           https://rpp-opendata.egon.gov.cz/odrpp/zdroj/kategorie-ovm/KO12      None                 None   
4          https://rpp-opendata.egon.gov.cz/odrpp/zdroj/kategorie-ovm/KO241      None                 None   
5           https://rpp-opendata.egon.gov.cz/odrpp/zdroj/kategorie-ovm/KO12      None                 None   
6           https://rpp-opendata.egon.gov.cz/odrpp/zdroj/kategorie-ovm/KO11      None                 None   
7          https://rpp-opendata.egon.gov.cz/odrpp/zdroj/kategorie-ovm/KO241      None                 None   

  kategorieId                kategorieNazev  
0        None                          None  
1        None                          None  
2        KO11  Obce s rozšířenou působností  
3        KO12          Kraje a hl. m. Praha  
4       KO241    Městské části Prahy 1 – 22  
5        KO12          Kraje a hl. m. Praha  
6        KO11  Obce s rozšířenou působností  
7       KO241    Městské části Prahy 1 – 22

#### JSON

V případě práce s JSON soubory je možné místní příslušnosti navázané k úkonu získat z JSON distribuce datové sady [Úkony](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F02925784ee989ac89fbb61d2e3b94c8e), kde je vazba mezi úkonem a místními příslušnostmi kódována v následující JSON struktuře (viz [dokumentace](https://data.gov.cz/otevřené-formální-normy/registr-práv-a-povinností/úkony/) pro jejich detailní popis):

```
{
  položky {
    místní-příslušnost {
      typ-subjektu,
      typ-místní-příslušnosti,
      poskytovatel-k-místní-příslušnosti
    }
  }
}
```

kde ```typ-subjektu``` je referencí na identifikátor typu subjektů uvedený v datové sadě [Číselník typů subjektů](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F281575084358cb8be628b7cc66020c2c), ```typ-místní-příslušnosti``` je referencí na identifikátor typu místní příslušnosti uvedený v datové sadě [Číselník typů místních příslušností k subjektům](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2Fcf79f7477c758122a8d99891e66a9800) a ```poskytovatel-k-místní-příslušnosti``` je referencí na identifikátor OVM z datové sady [Orgány veřejné moci](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F44a9d6abacd4d0e83a0694e74d028f51) nebo kategorie OVM z datové sady [Kategorie orgánů veřejné moci](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F4cdc3767e173806c2a7bb345459731a5).

### Sdružování úkonů podle věcných souvislostí - agendy

Pro aplikační využití dat z RPP by mohlo být užitečné sdružovat úkony dle nějaké věcné souvislosti. V případě, že se uživatel zajímá o nějaký úkon, mohla by aplikace nabízet související úkony. Bohužel neexistuje procesní model veřejné správy, který by mohl být využit pro navádění uživatele na následující či předcházející úkon v procesu, příp. na všechny úkony v daném procesu. Byly zpracovány tzv. životní situace, ty ale nejsou ve strojově zpracovatelné podobě a už vůbec ne navázány na úkony v RPP.

Jednu možnost sdružování nám RPP dává. Tou jsou agendy a činnosti, ze kterých se agendy skládají. Zatímco úkony jsou totiž pohledem na veřejnou správu z úhlu pohledu občana, podnikatele nebo právnické osoby, tak agendy a činnosti jsou pohledem na veřejnou správu z úhlu pohledu veřejné správy samotné, resp. legislativy, která veřejnou správu definuje.

Tak např. úkon U1073 "Zřízení nového válečného hrobu", který používáme jako příklad patří do agendy A549 "Péče o válečné hroby".

Takové sdružování podle agendy, může uživatelům pomoci dohledávat související úkony a navádět uživatele, aby na něco při vyřizování svých záležitostí nezapomněl. Dále může sdružování podle agendy, resp. doplnění údajů o úkonu údaji o agendě, vhodně doplnit informaci k úkonu, jehož název nemusí být zcela jasný. Např. z názvu úkonu U1094 nevyplývá, že je tento úkon spojen s válečnými hroby.

#### SPARQL a RDF

Následující dotaz vrací seznam všech úkonů, které do této agendy patří.

In [17]:
úkonyVAgendě=endpoint.select("""PREFIX l: <https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/>
PREFIX a: <https://slovník.gov.cz/agendový/104/pojem/>

SELECT ?ukon ?id ?nazev ?agenda
WHERE {
  VALUES (?agenda) {(<https://rpp-opendata.egon.gov.cz/odrpp/zdroj/agenda/A549>)}

  ?ukon l:je-vykonáván-v-agendě ?agenda ;
    l:má-identifikátor-úkonu ?id ;
    a:má-název-úkonu ?nazev .

}""")
úkonyVAgendě

ukon     id  \
0  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1072  U1072   
1  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
2  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1074  U1074   
3  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1075  U1075   
4  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1094  U1094   

                                                 nazev                                                    agenda  
0  Přemístění, zrušení nebo jiná změna válečného hrobu  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/agenda/A549  
1                       Zřízení nového válečného hrobu  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/agenda/A549  
2     Vydávání rozhodnutí o tom, co je válečným hrobem  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/agenda/A549  
3           Poskytování dotací na péči o válečné hroby  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/agenda/A549  
4               Náhrada za omezení užívání nemovitosti  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/agenda/A549

#### JSON

V případě JSON souborů lze vazbu na agendy získat v datové sadě [Úkony](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F02925784ee989ac89fbb61d2e3b94c8e), viz [popis výše](#JSON).

### Sdružování úkonů podle věcných souvislostí - činnosti

Agendy se skládají z činností, což je detailnější pohled na veřejnou správu z pohledu legislativy a veřejné správy samotné. Úkony nejsou vázány jen na agendy, jak bylo ukázáno výše, ale také na činnosti v těchto agendách. Vazba ale není tak přímočará. Pro činnosti je nutno odlišit poskytování úkonu a výkon úkonu popsané výše. Pro připomenutí, poskytování úkonu reflektuje to, jak je souvislost mezi úkonem a OVM zapsána v legislativě. Úkon může poskytovat OVM nebo celá kategorie OVM. Výkon úkonu pak značí, jak je to ve skutečnosti. Skutečnost se od legislativní definice může lišit např. z důvodu přenosu působností mezi obcemi apod.

Z pohledu poskytování úkonů jsou činnosti navázány na třídu [Poskytování úkonu](https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/poskytování-úkonu) prostřednictvím vazby [Je poskytován v rámci činnosti](https://slovník.gov.cz/agendový/104/pojem/je-poskytován-v-rámci-činnosti). Z pohledu výkonu úkonů jsou činnosti navázány na třídu [Poskytování úkonu](https://slovník.gov.cz/agendový/104/pojem/působnost-v-agendě) prostřednictvím vazby [Vykonává činnosti v rámci působnosti](https://slovník.gov.cz/agendový/104/pojem/vykonává-činnosti-v-rámci-působnosti). Konceptuální schéma je zachyceno na následujícím obrázku.

![Úkony a činnosti](img/úkony-činnosti.png)

Data o činnostech jsou dostupná v datové sadě [Agendy](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F9c73b802263c5e0ccf5542f10fbc35bb), která již byla zmiňována výše.

#### SPARQL

In [18]:
činnosti=endpoint.select("""PREFIX l: <https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/>
PREFIX a: <https://slovník.gov.cz/agendový/104/pojem/>

SELECT DISTINCT ?ukon ?id ?cinnost ?cinnostKod ?cinnostNazev
WHERE {
  VALUES (?ukon) {(<https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073>)}


  ?ukon l:má-identifikátor-úkonu ?id .

  {
    ?ukon l:je-poskytován/a:je-poskytován-v-rámci-činnosti ?cinnost .
  } UNION {
    ?pusobnost a:vykonává-úkon-v-rámci-působnosti/a:má-vykonávaný-úkon ?ukon ;
      a:vykonává-činnost-v-rámci-působnosti ?cinnost .
  }

  ?cinnost a:má-název-činnosti ?cinnostNazev ;
    a:má-kód-činnosti ?cinnostKod .
  
}""")
činnosti

ukon     id  \
0   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
1   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
2   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
3   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
4   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
5   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
6   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
7   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
8   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
9   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   
10  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1073  U1073   

                                                              cinnost cinnostKod  \
0    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/činnost/A549/CR1305     CR1305   
1   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/činnost/A549/CR10356    CR10356   
2    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/činnost/A549/CR1304     CR1304   
3    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/činnost/A549/CR1306     CR1306   
4    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/činnost/A549/CR1307     CR1307   
5    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/činnost/A549/CR1308     CR1308   
6    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/činnost/A549/CR1309     CR1309   
7    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/činnost/A549/CR1310     CR1310   
8    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/činnost/A549/CR1311     CR1311   
9    https://rpp-opendata.egon.gov.cz/odrpp/zdroj/činnost/A549/CR1312     CR1312   
10   https://rpp-opendata.egon.gov.cz/odrpp/zdroj/činnost/A549/CR1313     CR1313   

                                                                                           cinnostNazev  
0                                                            Souhlas se zřízením nového válečného hrobu  
1                                                            Projednávání přestupků a správních deliktů  
2   Souhlas s přemístěním nebo zrušením válečného hrobu anebo s jinou změnou válečného hrobu než je ...  
3                                                                       Vedení evidence válečných hrobů  
4                                                              Vedení souhrnné evidence válečných hrobů  
5                                                             Vedení centrální evidence válečných hrobů  
6                                                      Vydávání rozhodnutí o tom, co je válečným hrobem  
7                                       Koordinace péče o válečné hroby v České republice a v zahraničí  
8                                                         Koordinace péče o válečné hroby v rámci kraje  
9                                                            Poskytování dotací na péči o válečné hroby  
10                                                    Poskytování náhrad za omezení užívání nemovitosti

Opačně můžeme pro danou činnost zjistit seznam úkonů poskytovaných v rámci činnosti. Činnosti a úkony mohou odpovídat 1:1, ale nemusí. Následující dotaz vrací seznam úkonů poskytovaných v rámci činnosti CR1313 "Poskytování náhrad za omezení užívání nemovitosti".

In [19]:
úkonyPoskytovanéPřiČinnosti=endpoint.select("""PREFIX l: <https://slovník.gov.cz/legislativní/sbírka/111/2009/pojem/>
PREFIX a: <https://slovník.gov.cz/agendový/104/pojem/>

SELECT DISTINCT ?cinnost ?ukon ?id ?nazev
WHERE {
  VALUES (?cinnost) {(<https://rpp-opendata.egon.gov.cz/odrpp/zdroj/činnost/A549/CR1313>)}


  ?ukon l:je-poskytován/a:je-poskytován-v-rámci-činnosti ?cinnost ;
    l:má-identifikátor-úkonu ?id ;
    a:má-název-úkonu ?nazev .
  
}""")
úkonyPoskytovanéPřiČinnosti

cinnost  \
0  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/činnost/A549/CR1313   

                                                      ukon     id                                   nazev  
0  https://rpp-opendata.egon.gov.cz/odrpp/zdroj/úkon/U1094  U1094  Náhrada za omezení užívání nemovitosti

#### JSON

V případě práce s JSON soubory lze činnosti, v rámci kterých jsou úkony poskytovány získat z JSON distribuce datové sady [Úkony](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F02925784ee989ac89fbb61d2e3b94c8e), kde jsou kódovány v následující JSON struktuře vedle vazby na OVM či kategorie OVM poskytující úkon (viz [dokumentace](https://data.gov.cz/otevřené-formální-normy/registr-práv-a-povinností/úkony/) pro jejich detailní popis):

```
{
  položky {
    poskytování-úkonu {
      činnost
    }
  }
}
```

kde hodnotou prvku ```činnost``` je referencí na identifikátor činnosti.

Činnosti vykonávané v rámci působnosti v agendě, ve které jsou úkony vykonávány, lze získat z JSON distribuce datové sady [Působnost v agendách](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F7ee91a644bf6da16a9f5c4f337163c0f) (viz [dokumentace](https://data.gov.cz/otevřené-formální-normy/registr-práv-a-povinností/působnost-v-agendách/) pro detailní popis struktury). 

```
{
  položky {
    činnosti
  }
}
```

kde hodnotou prvku ```činnosti``` je seznam referencí na identifikátory činnosti.

Údaje o činnosti, jako např. výše uvedený název jsou potom v datové sadě [Agendy](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F9c73b802263c5e0ccf5542f10fbc35bb), kde jsou činnosti uvedeny v následující JSON struktuře:

```
{
  položky {
    činnosti {
      id,
      kód-činnosti,
      název-činnosti
    }
  }
}
```

kde prvek ```id``` je referencovaným identifikátorem z datové sady [Působnost v agendách](https://data.gov.cz/datov%C3%A1-sada?iri=https%3A%2F%2Fdata.gov.cz%2Fzdroj%2Fdatov%C3%A9-sady%2FMV%2F706529437%2F7ee91a644bf6da16a9f5c4f337163c0f).